File: java/148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            prev = prev.next;
            slow = slow.next;
            fast = fast.next;
            fast = fast.next;
        }
        this.print(prev.next);
        ListNode secondHead = sortList(prev.next);
        prev.next = null;
        ListNode firstHead = sortList(dummy.next);
        return merge(firstHead, secondHead);
    }
    private ListNode merge(ListNode first, ListNode second) {
        if (first == null && second == null) {
            return null;
        }
        if (first == null) {
            return second;
        }
        if (second == null) {
            return first;
        }
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;
        while (first != null && second != null) {
            if (first.val < second.val) {
                curr.next = first;
                first = first.next;
            } else {
                curr.next = second;
                second = second.next;
            }
            curr = curr.next;
        }
        while (first != null) {
            curr.next = first;
            first = first.next;
            curr = curr.next;
        }
        while (second != null) {
            curr.next = second;
            second = second.next;
            curr = curr.next;
        }
        curr.next = null;
        return dummy.next;
    }
}

File: python3/148.sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def printList(self, head):
        ans = ""
        while head != None:
            ans += str(head.val) + "->"
            head = head.next
        print(ans)
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        slow = head
        fast = head.next
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        newHead = slow.next
        slow.next = None
        head1 = self.sortList(head)
        head2 = self.sortList(newHead)
        dummy = ListNode(-1)
        curr = dummy
        pointerA = head1
        pointerB = head2
        while pointerA is not None and pointerB is not None:
            if pointerA.val < pointerB.val:
                curr.next = pointerA
                pointerA = pointerA.next
            else:
                curr.next = pointerB
                pointerB = pointerB.next
            curr = curr.next
        while pointerA is not None:
            curr.next = pointerA
            pointerA = pointerA.next
            curr = curr.next
        while pointerB is not None:
            curr.next = pointerB
            pointerB = pointerB.next
            curr = curr.next
        curr.next = None
        return dummy.next

        
        